Fallback to treating as object if not collection

ericstj opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/96652

The reflection binder will fallback if a type does not meet collection heuristics, but the source generator did not.

ericstj committed:
Fallback to treating as object if not collection

The reflection binder will fallback if a type does not meet collection
heuristics, but the source generator did not.

eiriktsarpalis commented on src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.cs:
```diff
@@ -1067,5 +1068,25 @@ public new TestSettingsEnum2 E
             public override int X { set => base.X = value + 1; }
         }
 
+        public class EnumerableNotCollection : IEnumerable<KeyValuePair<string, string>>
```
What specific attribute of this type makes it not a collection? Is it that it doesn't implement one of the known collection interfaces or that it's not mutable?

ericstj replied:
It doesn't match any of the heuristics for enum-like collections: https://github.com/dotnet/runtime/blob/dcf9814b9194cd1ab14ed8108098f06e0fd53e67/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs#L450-L540

In this particular case it's because it doesn't implement ICollection<>, nor expose an `Add` method: https://github.com/dotnet/runtime/blob/dcf9814b9194cd1ab14ed8108098f06e0fd53e67/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs#L485

eiriktsarpalis commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:
```diff
@@ -198,6 +198,12 @@ private TypeSpec CreateTypeSpec(TypeParseInfo typeParseInfo)
                     spec = CreateCollectionSpec(typeParseInfo);
+
+                    // fallback to treating as an object, if we can create it
+                    if (spec is UnsupportedTypeSpec && type is INamedTypeSymbol && !type.IsAbstract)
```
What types of inputs are deemed to be collections but then return `UnsupportedTypeSpec`? Could these checks be integrated in the `IsCollection` predicate instead?

ericstj replied:
I considered that  -- it's all of the checks here: https://github.com/dotnet/runtime/blob/dcf9814b9194cd1ab14ed8108098f06e0fd53e67/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs#L363-L540

IMO pulling those out into the checking method would create two codepaths we have to keep in sync, since many of these checks would need to be duplicated when examining the type and deciding how it would bind.

eiriktsarpalis replied:
Could they be merged into a single `Try-` method that just `out`s a valid model? Just a suggestion, I realize we have a deadline today and time is short.

ericstj replied:
It's actually a bit more nuanced than just returning false here.  Much of the time we want the diagnostics info from this.  The fallback turns out to be a bit simpler than this.

It's if we didn't find a way to bind as a collection, and the type can be created and does not implement ICollection<> nor IDictionary<,> than treat as an object.

ericstj committed:
Fix UnsupportedTypes test

ericstj commented:
So this has some slightly unusual behavior.   The unsupported types test changed as follows.

`List<nint>` is now supported and binds only the `Capacity` member.

A derived `Dictionary<string,Action>` is now supported and tries to bind the `Keys` but fails to actually do anything.

While this is unusual I'm not sure if we care.  Probably the most undesirable part of this is that it used to emit:
```
warning SYSLIB1100: The collection type is not supported: 'MyDictionary'. (https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1100)
 warning SYSLIB1100: The collection type is not supported: 'System.Collections.Generic.List<nint>'. (https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1100)
 ```
 
And now it emits:
```
warning SYSLIB1100: Cannot create instance of type 'System.Collections.Generic.IEqualityComparer<string>' because it is missing a public instance constructor. (https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1100)
warning SYSLIB1101: Property 'Comparer' on type 'MyDictionary' is not supported. (https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1101)
warning SYSLIB1100: Cannot create instance of type 'System.Collections.Generic.Dictionary<string, Action>+KeyCollection' because it is missing a public instance constructor. (https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1100)
warning SYSLIB1101: Property 'Keys' on type 'MyDictionary' is not supported. (https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1101)
warning SYSLIB1100: The collection type is not supported: 'System.Collections.ICollection'. (https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1100)
warning SYSLIB1101: Property 'System.Collections.IDictionary.Keys' on type 'MyDictionary' is not supported. (https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1101)
```

This is different than the reflection binder.  The reflection binder only falls back to object binding if the type does not implement IDictionary<> or ICollection<>.  I think I'll try @eiriktsarpalis's [suggestion](https://github.com/dotnet/runtime/pull/106378#discussion_r1717128328).

ericstj committed:
Update collection to object fallback condition

This matches what the refelction binder does, and fixes the baseline
diffs (and diagnostics changes) we were seeing for unsupported
collection types.

ericstj commented:
Ok, I fixed that and it addressed the diff in baselines/diagnostics.  This is ready for a re-review.

eiriktsarpalis commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:
```diff
@@ -198,6 +198,15 @@ private TypeSpec CreateTypeSpec(TypeParseInfo typeParseInfo)
+                    // fallback to treating as an object if we couldn't treat as a collection and
+                    // it does not implement ICollection<> nor IDictionary<,>
+                    if (spec is UnsupportedTypeSpec && type is INamedTypeSymbol namedType &&
+                        GetInterface(namedType, _typeSymbols.GenericICollection_Unbound) is null &&
```
Do these interfaces also get resolved inside the `CreateCollectionSpec` method? If so, would it be worth avoiding additional a bit of duplicated work by returning an `out bool doesNotImplementKnownMutableCollectionInterfaces` parameter? Or perhaps embed that information in the `UnsupportedTypeSpec` for a cleaner look?

ericstj committed:
Refactor fallback from collection to object mode

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:
```diff
@@ -360,34 +360,37 @@ private TypeSpec CreateArraySpec(TypeParseInfo typeParseInfo)
                 if (IsUnsupportedType(keyTypeSymbol) || IsUnsupportedType(elementTypeSymbol))
                 {
-                    return CreateUnsupportedCollectionSpec(typeParseInfo);
+                    return isDictionary ? CreateUnsupportedCollectionSpec(typeParseInfo) : null;
```
Can't this type is implementing a non-generic IDictionary?

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:
```diff
@@ -402,14 +405,15 @@ private TypeSpec CreateDictionarySpec(TypeParseInfo typeParseInfo, ITypeSymbol k
                     else
                     {
-                        return CreateUnsupportedCollectionSpec(typeParseInfo);
+                        // not a dictionary
```
Can't this type is implementing a non-generic IDictionary?

ericstj replied:
It could, but we don't treat those as bindable dictionaries.
https://github.com/dotnet/runtime/blob/2106bb0c57c7bc39c9b5320668c09321651b99e6/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L447-L452

I'm not thrilled with the naming of this local, but it was the best I could come up with.  Here the "not a dictionary" means we're going to treat the return value as not a dictionary so we should fallback to object binding behavior.

tarekgh commented on src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:
```diff
@@ -475,14 +481,15 @@ private TypeSpec CreateEnumerableSpec(TypeParseInfo typeParseInfo)
                     {
-                        return CreateUnsupportedCollectionSpec(typeParseInfo);
+                        // not a collection
+                        return null;
```
>return null; [](http://example.com/codeflow?start=24&length=12)

Can't this type is implementing a non-generic ICollection?

ericstj replied:
Same as above.  It could, but we don't treat those as bindable collections.
https://github.com/dotnet/runtime/blob/2106bb0c57c7bc39c9b5320668c09321651b99e6/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L455-L459

ericstj commented:
/backport to release/9.0-rc1
